### PR TITLE
Calcul des AL dans les DOMs & gros refactoring

### PR DESCRIPTION
**Pull request by fpagnoux**
Connected to sgmap/openfisca-france#50
Depends on https://github.com/openfisca/openfisca-core/pull/385

I split the `aide_logement_montant_brut variable`, it was way too big and a nightmare to debug. So a lot of new variables are introduced, but I think the complexity justifies it.

### CODE REVIEW

**Comment by benjello** [comment]
Whu don't you use helper fonction to compute specific portions. You will avoid creating intermediate variable which will be very difficult to handle with survey data. You can see a simple example [here] (https://github.com/openfisca/openfisca-france/blob/master/openfisca_france/model/prelevements_obligatoires/prelevements_sociaux/cotisations_sociales/allegements.py#L411)

**Comment by fpagnoux** [comment]
I'm resuscitating this PR.
It's been in prod for some time in mes-aides, and it was tedious to update it with current openfisca. We cannot let it rotten for much longer.

I don't think anyone had time to work on the cache optimization though... @benjello is it ok for you to merge this anyway ?

To make progress the cache optimization, imo we need to:
- [ ] Build tests and some doc about the parsers so that they are usable by contributors.
- [ ] Improve the parsers to detect more reliably the input variables of a formula.
With this, we should be able to compute the tree. Then:
- [ ] Design an algorithm to determine which variables to put in cache when using the "memory saving" mode. We can probably use a very basic one, but we need to check the efficiency empirically.
- [ ] Implement the flexible cache in openfisca-core.

That seems to be a lot of work...

**Comment by benjello** [comment]
Since it will be a lot of work and it may break the computation with data, merging this kind of PR put me at risk of not being able to run with data when needed ...
And I do not see a lot of incentives for you to build the cache ;-)
So I am worried.
Please convince me that I can have a fall back if I need some memory.

**Comment by fpagnoux** [comment]
A fallback, more simple but not durable, could be to have a blacklist of variables we do not want to put in cache.

**Comment by benjello** [comment]
Yes indeed ! I can live with that.

**Comment by fpagnoux** [comment]
I implemented an opt out mechanism for the cache. See the [PR](openfisca/openfisca-core#385) on `core`.

To use it for `openfisca_france`, just add the variables you want to exclude from the cache [here](https://github.com/sgmap/openfisca-france/blob/7a231d5d335da731d9ecadf2fc90a48c91fce2ee/openfisca_france/conf/cache_blacklist.py) and, when creating the simulation, add an extra parameter:
```python
simulation = scenario.new_simulation(opt_out_cache = True)
```
See the [test file](https://github.com/sgmap/openfisca-france/blob/7a231d5d335da731d9ecadf2fc90a48c91fce2ee/openfisca_france/tests/test_opt_out_cache.py) for a basic example.

**Comment by benjello** [comment]
This is excellent and should be sufficient to avoid out of memory errors.
No idea about where to put your conf file.

File: openfisca_france/conf/cache_blacklist.py [hunk 1]
```diff
@@ -2,4 +2,10 @@
 # The following variables are intermadiate results and do not need to be cached in those usecases.
```
Comment (laem) [inline openfisca_france/conf/cache_blacklist.py]: intermadiate -> intermediate

**Pull request merged**